EMULATOR/LIBGPU.C

```diff
@@ -470,92 +529,148 @@ void DrawSplit(const VertexBufferSplit &split)
-
 		P_TAG* pTag = (P_TAG*)p;
 
-		do
+		// P_TAG as primitive list
+		//do
+		while ((uintptr_t)pTag != (uintptr_t)&terminator)
 		{
 			if (pTag->len > 0)
 			{
-				ParseLinkedPrimitiveList((uintptr_t)pTag, (uintptr_t)pTag + (uintptr_t)(pTag->len * 4) + 4 + LEN_OFFSET);
+				int lastSize = ParseLinkedPrimitiveList((uintptr_t)pTag, (uintptr_t)pTag + (uintptr_t)(pTag->len * 4) + 4 + LEN_OFFSET);
+				if (lastSize == -1)
+					break; // safe bailout
 			}
 			pTag = (P_TAG*)pTag->addr;
-		}while ((uintptr_t)pTag != (uintptr_t)&terminator);
-
-		Emulator_UpdateVertexBuffer(g_vertexBuffer, g_vertexIndex);
-
-		for (int i = 1; i <= g_splitIndex; i++)
-		{
-			DrawSplit(g_splits[i]);
 		}
 	}
+}
 
-#if defined(PGXP)
-	/* Reset the ztable */
-	memset(&pgxp_vertex_buffer[0], 0, pgxp_vertex_index * sizeof(PGXPVertex));
+//------------------------------------------------------------------
 
-	/* Reset the ztable index of */
-	pgxp_vertex_index = 0;
+void DrawOTagEnv(u_long* p, DRAWENV* env)
+{
+	PutDrawEnv(env);
+	DrawOTag(p);
+}
+
+void DrawOTag(u_long* p)
+{
+	if (Emulator_BeginScene())
+	{
+		ClearVBO();
+		ResetPolyState();
+	}
+
+#if defined(DEBUG_POLY_COUNT)
+	polygon_count = 0;
 #endif
-	Emulator_EndScene();
+
+	if (activeDrawEnv.isbg)
+	{
+		ClearImage(&activeDrawEnv.clip, activeDrawEnv.r0, activeDrawEnv.g0, activeDrawEnv.b0);
+	}
+	else
+	{
+		Emulator_BlitVRAM();
+	}
+
+	AggregatePTAGsToSplits(p, false);
+
+	DrawAggregatedSplits();
+	//Emulator_EndScene();
 }
 
-void ParseLinkedPrimitiveList(unsigned int packetStart, unsigned int packetEnd)//@TODO sync with ParsePrimitive
+void DrawPrim(void* p)
 {
-	unsigned int currentAddress = packetStart;
+	if (Emulator_BeginScene())
+	{
+		ClearVBO();
+		ResetPolyState();
+	}
 
-	while (currentAddress != packetEnd)
+#if defined(DEBUG_POLY_COUNT)
+	polygon_count = 0;
+#endif
+
+	if (activeDrawEnv.isbg)
 	{
-		P_TAG* pTag = (P_TAG*)currentAddress;
+		ClearImage(&activeDrawEnv.clip, activeDrawEnv.r0, activeDrawEnv.g0, activeDrawEnv.b0);
+	}
+	else {
+		Emulator_BlitVRAM();
+	}
+
+	AggregatePTAGsToSplits((u_long*)p, true);
+}
+
+// parses primitive and pushes it to VBO
+// returns primitive size
+// -1 means invalid primitive
+int ParsePrimitive(uintptr_t primPtr)
+{
+	P_TAG* pTag = (P_TAG*)primPtr;
 
-		int textured = (pTag->code & 0x4) != 0;
+	int textured = (pTag->code & 0x4) != 0;
 
-		int blend_mode = 0;
+	int blend_mode = 0;
 
-		if (textured)
+	if (textured)
+	{
+		if ((pTag->code & 0x1) != 0)
 		{
-			if ((pTag->code & 0x1) != 0)
-			{
-				blend_mode = 2;
-			}
-			else
-			{
-				blend_mode = 1;
-			}
+			blend_mode = 2;
 		}
 		else
 		{
-			blend_mode = 0;
+			blend_mode = 1;
 		}
+	}
+	else
+	{
+		blend_mode = 0;
+	}
 
-		bool semi_transparent = (pTag->code & 2) != 0;
+	bool semi_transparent = (pTag->code & 2) != 0;
 
-		switch (pTag->code & ~3)
-		{
+	int primitive_size = -1;	// -1
+
+	switch (pTag->code & ~3)
+	{
 		case 0x0:
 		{
-			currentAddress += 4;
+			primitive_size = 4;
```

Comment:
Why explicitly set the primitive size? What was wrong with using currentAddress? Now it's more tricky to debug problems.

Reply:
It's because ParsePrimitive returns only primitive size so ParsePrimitiveLinkedList loop can advance by it.